### PR TITLE
[HTTP CLIENT] Add RequestEvent once a request is send

### LINKED ISSUE #48: Dispatch an Event
Dispatch an event when trigger a request

### PR DESCRIPTION
**Pull request by ismail1432**
Closes #48 

Dispatch event when a Request is triggered
- [x] Dispatch
- [x] Tests

### CODE REVIEW

File: src/HttpClient/GenericHttpClient.php [hunk 1]
```diff
@@ -32,11 +35,13 @@ class GenericHttpClient implements ReplaceableHttpClientInterface
     public function __construct(
         SerializerInterface $serializer,
         HttpClientInterface $httpClient,
-        iterable $authenticationHeaderProviders = []
+        iterable $authenticationHeaderProviders = [],
+        EventDispatcherInterface $dispatcher = null
```
Comment (mtarld) [inline src/HttpClient/GenericHttpClient.php:39]:
```suggestion
        ?EventDispatcherInterface $dispatcher = null
```

File: src/HttpClient/GenericHttpClient.php [hunk 2]
```diff
@@ -73,7 +78,13 @@ public function request(Microservice $microservice, string $method, string $uri,
             }
         }
 
-        return $this->httpClient->request($method, $uri, $options);
+        $response = $this->httpClient->request($method, $uri, $options);
```
Comment (mtarld) [inline src/HttpClient/GenericHttpClient.php:81]: Here, `HttpClientInterface::request` can throw a `TransportExceptionInterface`.
But if that exception is thrown, we won't dispatch the `RequestEvent`.
Maybe we can surround it with a try/catch block a dispatch that event in the finally? (with a nullable `Response` then)

File: src/Event/RequestEvent.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
+<?php
+
+namespace Mtarld\ApiPlatformMsBundle\Event;
+
+use Mtarld\ApiPlatformMsBundle\Microservice\Microservice;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * @final
```
Comment (mtarld) [inline src/Event/RequestEvent.php:9]: Could you add `psalm-immutable`?

File: CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,9 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [Unreleased]
+* Added the `RequestEvent` dispatched once a request is send
```
Comment (mtarld) [inline CHANGELOG.md:9]:
```suggestion
* Added dispatching of a `RequestEvent` once a request is sent
```

File: tests/Event/RequestEventTest.php [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * @author Smaïne Milianni <smaine.milianni@gmail.com>
+ */
+class RequestEventTest extends TestCase
+{
+    public function testEventIsDispatch()
```
Comment (mtarld) [inline tests/Event/RequestEventTest.php:19]: This test should be in `HttpClientTest` IMHO

Comment (mtarld) [inline tests/Event/RequestEventTest.php:19]:
```suggestion
    public function testEventIsDispatched(): void
```

File: tests/Event/RequestEventTest.php [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    {
+        $httpClient = new GenericHttpClient(
+            $this->createMock(SerializerInterface::class),
+            $this->createMock(HttpClientInterface::class),
+            [],
+            $eventDispatcherMock = $this->createMock(EventDispatcherInterface::class)
```
Comment (mtarld) [inline tests/Event/RequestEventTest.php:25]:
```suggestion
            $eventDispatcherSpy = $this->createMock(EventDispatcherInterface::class)
```

File: tests/Event/RequestEventTest.php [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        );
+
+        $microservice = $this->createMock(Microservice::class);
+        $microservice->method('getFormat')->willReturn('jsonld');
+
+        $eventDispatcherMock->expects($this->once())->method('dispatch');
```
Comment (mtarld) [inline tests/Event/RequestEventTest.php:31]: You can as well use `with(...arguments)` to verify that the dispatched event is the expected one.

File: tests/Event/RequestEventTest.php [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+        $httpClient->request($microservice, 'GET', '/foo');
+    }
+
+    /**
+     * @dataProvider requestProvider
+     */
+    public function testGetter(Microservice $microservice, string $method, string $uri, array $options, ResponseInterface $response, string $content)
```
Comment (mtarld) [inline tests/Event/RequestEventTest.php:39]: IMHO, this test is not needed as `RequestEvent` is an immutable POPO

File: src/HttpClient/GenericHttpClient.php [hunk 9]
```diff
@@ -73,7 +78,13 @@ public function request(Microservice $microservice, string $method, string $uri,
...
+
+        if (null !== $this->dispatcher) {
+            $this->dispatcher->dispatch(new RequestEvent($microservice, $method, $uri, $options, $response));
```
Comment (mtarld) [inline src/HttpClient/GenericHttpClient.php:84]: Could you add `RequestEvent` to preloading?

```php
// Help opcache.preload discover always-needed symbols
class_exists(RequestEvent::class);
```

**Review by mtarld — CHANGES_REQUESTED** [review]

File: src/HttpClient/GenericHttpClient.php [hunk 10]
```diff
@@ -73,7 +78,19 @@ public function request(Microservice $microservice, string $method, string $uri,
             }
         }
 
-        return $this->httpClient->request($method, $uri, $options);
+        try {
+            $response = $this->httpClient->request($method, $uri, $options);
+        } catch (ExceptionInterface $exception) {
+            $response = $exception->getResponse();
```
Comment (mtarld) [inline src/HttpClient/GenericHttpClient.php:84]: `getResponse` is part of `HttpExceptionInterface` and not `ExceptionInterface`.
Therefore, you should test if the exception is a `HttpExceptionInterface` (using `instanceof`?) and set the response only if it is

File: src/HttpClient/GenericHttpClient.php [hunk 11]
```diff
@@ -73,7 +78,19 @@ public function request(Microservice $microservice, string $method, string $uri,
...
+
+            throw $exception;
+        } finally {
+            if (null !== $this->dispatcher) {
+                $this->dispatcher->dispatch(new RequestEvent($microservice, $method, $uri, $options, $response));
```
Comment (mtarld) [inline src/HttpClient/GenericHttpClient.php:89]:
```suggestion
                $this->dispatcher->dispatch(new RequestEvent($microservice, $method, $uri, $options, $response ?? null));
```

File: tests/HttpClient/HttpClientTest.php [hunk 12]
```diff
@@ -162,10 +165,47 @@ public function testAuthenticationHeaders(string $microservice, string $method,
...
+
+        $microservice = $this->createMock(Microservice::class);
+        $microservice->method('getFormat')->willReturn($format);
+        $microservice->method('getBaseUri')->willReturn($uri);
+
+        $response = $httpClient->request($microservice, $method, $uri);
+
+        $eventDispatcherSpy->expects($this->once())->method('dispatch')->with(new RequestEvent($microservice, $method, $uri, $options, $response));
+
+        $httpClient->request($microservice, $method, $uri);
```
Comment (mtarld) [inline tests/HttpClient/HttpClientTest.php:196]: Is this line needed?

File: composer.json [hunk 13]
```diff
@@ -24,7 +24,9 @@
     "symfony/property-access": "^4.4|^5.0",
     "symfony/serializer": "^4.4|^5.0",
     "symfony/validator": "^4.4|^5.0",
-    "symfony/polyfill-php80": "^1.20"
+    "symfony/event-dispatcher": "^4.4|^5.0",
+    "symfony/polyfill-php80": "^1.20",
+    "psr/log": "^1.0"
```
Comment (mtarld) [inline composer.json:29]: I don't this is part of the current PR ;)

File: tests/HttpClient/HttpClientTest.php [hunk 14]
```diff
@@ -162,10 +165,47 @@ public function testAuthenticationHeaders(string $microservice, string $method,
         static::$container->get(GenericHttpClient::class)->request(static::$container->get(MicroservicePool::class)->get($microservice), $method, $uri);
     }
 
+    /**
+     * @dataProvider eventProvider
+     */
+    public function testEventIsDispatched(string $method, string $uri, string $format, string $header): void
```
Comment (mtarld) [inline tests/HttpClient/HttpClientTest.php:171]: Maybe we can find a more meaningful name for `$header`? Something like `$mimeType`?

**Review by mtarld — CHANGES_REQUESTED** [review]
Almost good :wink:

File: composer.json (hunk 13 above)
Comment (ismail1432) [reply composer.json:29]: wrong cherry-pick :stuck_out_tongue:

File: src/Event/RequestEvent.php [hunk 15]
```diff
@@ -19,7 +19,7 @@ class RequestEvent
     private $options;
     private $response;
 
-    public function __construct(Microservice $microservice, string $method, string $uri, array $options, ResponseInterface $response)
+    public function __construct(Microservice $microservice, string $method, string $uri, array $options, ?ResponseInterface $response = null)
```
Comment (mtarld) [inline src/Event/RequestEvent.php:22]: I'd rather write `?ResponseInterface $response)` because I don't see the benefit to set a default value here

**Review by mtarld — CHANGES_REQUESTED** [review]

File: src/Event/RequestEvent.php (hunk 15 above)
Comment (ismail1432) [reply src/Event/RequestEvent.php:22]: good point

**Review by mtarld — APPROVED** [review]
LGTM! Many thanks!

**Pull request merged**